core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +117,93 @@ sealed trait Channel[F[_], A] {
+          )
+
+        def trySend(a: A): F[Either[Channel.Closed, Boolean]] =
+          isClosed.ifM(Channel.Closed.asLeft[Boolean].pure[F], q.tryOffer(a).map(Right(_)))
+
+        def stream: Stream[F, A] = {
+          val takeN: F[Chunk[A]] =
+            q.tryTakeN(None).flatMap {
+              case Nil =>
+                val fallback = leasesDrained.flatMap { b =>
+                  if (b) {
+                    MonadCancel[F].uncancelable { poll =>
+                      poll(Spawn[F].racePair(q.take, closedR.get)).flatMap {
+                        case Left((oca, fiber)) =>
+                          oca.embedNever.flatMap(a => fiber.cancel.as(Chunk.singleton(a)))
+
+                        case Right((fiber, ocb)) =>
+                          ocb.embedNever.flatMap { _ =>
+                            (fiber.cancel *> fiber.join).flatMap { oca =>
+                              oca.fold(
+                                Chunk.empty[A].pure[F],
+                                _ => Chunk.empty[A].pure[F],
+                                _.map(Chunk.singleton(_))
+                              )
+                            }
+                          }
                       }
-
-                      val toEmit = makeChunk(initValues, tailValues.result(), size)
-
-                      unblock.as(Pull.output(toEmit) >> consumeLoop)
-                    } else {
-                      F.pure(
-                        if (closed) Pull.done
-                        else Pull.eval(waiting.get) >> consumeLoop
-                      )
                     }
+                  } else {
+                    q.take.map(Chunk.singleton(_))
+                  }
                 }
-                .uncancelable
-            }
-          }.flatten
 
-        def notifyStream(waitForChanges: Option[Deferred[F, Unit]]) =
-          waitForChanges.traverse(_.complete(()))
+                isQuiesced.ifM(Chunk.empty[A].pure[F], fallback)
 
-        def waitOnBound(producer: Deferred[F, Unit], poll: Poll[F]) =
-          poll(producer.get).onCancel {
-            state.update { s =>
-              s.copy(producers = s.producers.filter(_._2 ne producer))
+              case as =>
+                Chunk.seq(as).pure[F]
             }
-          }
-
-        def signalClosure = closedGate.complete(())
-
-        @inline private def shouldEmit(s: State) = s.values.nonEmpty || s.producers.nonEmpty
 
-        private def makeChunk(init: List[A], tail: List[A], size: Int): Chunk[A] = {
-          val arr = new Array[Any](size)
-          var i = size - 1
-          var values = tail
-          while (i >= 0) {
-            if (values.isEmpty) values = init
-            arr(i) = values.head
-            values = values.tail
-            i -= 1
-          }
-          Chunk.array(arr).asInstanceOf[Chunk[A]]
+          // you can do this more efficiently, just proves a point
+          Stream.eval(takeN).repeat.takeWhile(!_.isEmpty).flatMap(Stream.chunk(_))
```

Comment:
lol?

Reply:
```suggestion
          // you can do this more efficiently, just proves a point
          Stream.eval(takeN).repeat.takeWhile(!_.isEmpty).unchunks
```

Reply:
Oh nice. Is that new?

Reply:
It was one of my first PRs to FS2 :)
- https://github.com/typelevel/fs2/pull/2506

Reply:
I've honestly wanted that convenience function for so long. I spend my whole life doing it.